modules/database/test/std/rec/biTest.c

```diff
@@ -0,0 +1,115 @@
+/*************************************************************************\
+* Copyright (c) 2023 Karl Vestin
+* EPICS BASE is distributed subject to a Software License Agreement found
+* in file LICENSE that is included with this distribution.
+\*************************************************************************/
+
+#include <unistd.h>
+
+#include "dbUnitTest.h"
+#include "testMain.h"
+#include "errlog.h"
+#include "dbAccess.h"
+#include "menuAlarmSevr.h"
+
+void recTestIoc_registerRecordDeviceDriver(struct dbBase *);
+
+static void test_soft_input(void){
+    /* set soft channel */
+    testdbPutFieldOk("test_bi_rec.DTYP", DBF_STRING, "Soft Channel");
+    testdbPutFieldOk("test_bi_rec.INP", DBF_STRING, "test_bi_link_rec.VAL");
+    testdbPutFieldOk("test_bi_link_rec.FLNK", DBF_STRING, "test_bi_rec");
+
+    /* set VAL to on linked record */
+    testdbPutFieldOk("test_bi_link_rec.VAL", DBF_SHORT, TRUE);
+
+    /* verify that this record VAL is updated but RVAL is not */
+    testdbGetFieldEqual("test_bi_rec.VAL", DBF_SHORT, TRUE);
+    testdbGetFieldEqual("test_bi_rec.RVAL", DBF_SHORT, FALSE);
+
+    // number of tests = 6
+}
+
+static void test_raw_soft_input(void){
+    /* set soft channel */
+    testdbPutFieldOk("test_bi_rec.DTYP", DBF_STRING, "Raw Soft Channel");
+    testdbPutFieldOk("test_bi_rec.INP", DBF_STRING, "test_bi_link_rec.VAL");
+    testdbPutFieldOk("test_bi_link_rec.FLNK", DBF_STRING, "test_bi_rec");
+
+    /* set VAL to on linked record */
+    testdbPutFieldOk("test_bi_link_rec.VAL", DBF_SHORT, TRUE);
+
+    /* verify that this record RVAL and VAL are updated */
+    testdbGetFieldEqual("test_bi_rec.VAL", DBF_SHORT, TRUE);
+    testdbGetFieldEqual("test_bi_rec.RVAL", DBF_SHORT, TRUE);
+
+    // number of tests = 6
+}
+
+static void test_operator_display(void){
+    /* set operator display parameters */
+    testdbPutFieldOk("test_bi_rec.ZNAM", DBF_STRING, "ZNAM_TEST");
+    testdbPutFieldOk("test_bi_rec.ONAM", DBF_STRING, "ONAM_TEST");
+    testdbPutFieldOk("test_bi_rec.DESC", DBF_STRING, "DESC_TEST");
+
+    /* verify operator display parameters */
+    testdbGetFieldEqual("test_bi_rec.ZNAM", DBF_STRING, "ZNAM_TEST");   
+    testdbGetFieldEqual("test_bi_rec.ONAM", DBF_STRING, "ONAM_TEST"); 
+    testdbGetFieldEqual("test_bi_rec.NAME", DBF_STRING, "test_bi_rec");
+    testdbGetFieldEqual("test_bi_rec.DESC", DBF_STRING, "DESC_TEST");
+
+    // number of tests = 7
+}
+
+static void test_alarm(void){
+    /* set soft channel */
+    testdbPutFieldOk("test_bi_rec.DTYP", DBF_STRING, "Soft Channel");
```

Comment:
If every test is a `Soft Channel`, maybe this should just be included in the `biTest.db` file.

Reply:
But they are not, there's a specific test for `Raw Soft Channel`. Arguably, the `DTYP` could be overriden just in that test, but I don't think it's any better for clarity.

Reply:
Ah, I missed that. I think you're right about clarity in that case.

Reply:
I actually prefer to keep the test code in the C file for readability, even if it does get a bit repetitive. Also more flexible for adding tests that may want a raw soft channel for instance.